### ISSUE TITLE
ngramDistance performance

### ISSUE THREAD

**Issue by SharonCastel** [ClickHouse/ClickHouse#39754]
Hi Guys!
A long time query execution case of a 700k rows table, with text column which analyzes similarity inside this table and use self join and executes ngramdistance on all the possible combination(one side).
make a long story short... the query itself takes 5.5 hours but when executing the function it goes up to 13.5 hours:
read_rows	read_bytes
1465876		130666038

``` select
        first_customer_number,
        second_customer_number,
        first_item,
        second_item,
        first_strip_name,
        second_strip_name,
        case when (length(first_strip_name) < 4 or length(second_strip_name) < 4) then
            0
            when (first_strip_name=second_strip_name) THEN 
            1
            else
                1 - ngramDistance(first_strip_name, second_strip_name)
        end as score,
        'name' as rule_name
    from (select customer_number as first_customer_number, 
    			 name_1 as first_item, 
    			 trim(replaceRegexpAll(lower(name_1), '[^\w]+', ' ')) as first_strip_name
    		from customers
    		where length(first_strip_name)>3) as oc1, 
	     (select customer_number as second_customer_number, 
				 name_1 as second_item, 
				 trim(replaceRegexpAll(lower(name_1), '[^\w]+', ' ')) as second_strip_name 
	     	from customers
    		where length(second_strip_name)>3) as oc2
    where first_customer_number > second_customer_number
        and score > 0.85

hope to have some performance advice here please🙌

**Comment by alexey-milovidov** [comment]
How many records are being processed and how long are the strings?
As far as I remember, the performance of ngramDistance function is just as good as it can be.

**Comment by den-crane** [comment]
As I understand ngramDistance is called 490000000000 times

**Comment by SharonCastel** [comment]
```
select histogram(5)(length(name_1))  from customers
['(0','25.767424058506585','637947.125)','(25.767424058506585','39.181284070260475','91555.375)','(39.181284070260475','55.769781722842644','2531)','(55.769781722842644','79.42211109796031','747.125)','(79.42211109796031','105','157.375)']
```
So most of the records with strings in size of 4-25 chars.
@alexey-milovidov this is for you: https://www.linkedin.com/posts/activity-6958106044547932160-Foeq?utm_source=linkedin_share&utm_medium=member_desktop_web

Yes Den you are right its the combination of all pairs without duplicates.
for now I succeeded to filter half of the records so the time got dramatically decreased.
In one query I succeeded to do the same as a full python code we had.
Still I wish the function was a bit optimized.
